File: exponentation_by_squaring/exponentation_by_squaring.c
#include<stdio.h>

/* Algorithm to calculate x raised to power k very efficiently
*  Can be used as a great alternative to normal pow function
*  base - base of the expression
*  power - power of expression
*/

long int exponentation_by_squaring(int base, int power) {
	if (power == 0)		// Base case1
		return 1;
	else if (power == 1)	// Base case2
		return base;
	else if (power % 2 == 0)
		return exponentation_by_squaring(base * base, power / 2);
	else if (power % 2 == 1)
		return exponentation_by_squaring(base * base, (power - 1) / 2);
}

int main() {
	int base = 2;
	int power = 30;
	printf("%d raised to %d is %ld\n", base, power, exponentation_by_squaring(base, power));
	return 0;
}

File: exponentation_by_squaring/exponentation_by_squaring.py
def exponentation_by_squaring(base, power):
    """
    Calculates base ^ power using recursion, efficient algorithm to inbuilt pow() function
    :param base: Base of expression
    :param power: Power of expression
    :return: returns base ^ power
    """

    if power < 0:  # Negative powers
        return exponentation_by_squaring(1. / base, -power)
    elif power == 0:  # Base case
        return 1
    elif power % 2 == 0:
        return exponentation_by_squaring(base * base, power // 2)
    elif power % 2 == 1:
        return base * exponentation_by_squaring(base * base, (power - 1) // 2)


def main():
    """
    driver function to calculate base ^ power
    """

    base = 2
    power = 3
    print(str(base) + ' raised to ' + str(power) + ' is ' +
          str(exponentation_by_squaring(base, power)))


if __name__ == '__main__':
    main()
